### PR TITLE
Fix a race in pod backoff.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/kubernetes/kubernetes/issues/14449

@lavalamp @quinton-hoole @erictune @smarterclayton

### CODE REVIEW

**Comment by lavalamp** [comment]
I doubt this is the right fix. it implies we're testing something about the pod twice simultaneously, which seems quite wrong in a case where we want to back off on repeated accesses?

**Comment by brendandburns** [comment]
I see multiple other issues in this codepath.  Will work w/ @lavalamp and extend this PR.

**Comment by brendandburns** [comment]
Ok, updated per our discussions, ptal.

thanks!
--brendan

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -274,9 +276,16 @@ func (factory *ConfigFactory) makeDefaultErrorFunc(backoff *podBackoff, podQueue
...
+			podID := pod.Namespace + "/" + pod.Name
 			podNamespace := pod.Namespace
-			backoff.wait(podID)
+
+			entry := backoff.getEntry(podID)
+			if !entry.TryLock() {
+				glog.Warningf("Request for pod %v already in flight, abandoning", podID)
+				return
+			}
+			defer entry.Unlock()
```
Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: Combine the trylock/wait/unlock group into a separate function. Unlock *must* happen before podQueue.Add, or you could falsely block a subsequent error handler under really pessimistic thread ordering.

File: plugin/pkg/scheduler/factory/factory.go [hunk 2]
```diff
@@ -335,8 +344,19 @@ func (realClock) Now() time.Time {
 }
 
 type backoffEntry struct {
-	backoff    time.Duration
-	lastUpdate time.Time
+	backoff     int64
```
Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: backoff can go back to being a time.Duration, right?

File: plugin/pkg/scheduler/factory/factory.go [hunk 3]
```diff
@@ -274,9 +276,16 @@ func (factory *ConfigFactory) makeDefaultErrorFunc(backoff *podBackoff, podQueue
...
+			entry := backoff.getEntry(podID)
+			if !entry.TryLock() {
+				glog.Warningf("Request for pod %v already in flight, abandoning", podID)
+				return
+			}
+			defer entry.Unlock()
+			entry.wait(backoff.maxDuration)
 			// Get the pod again; it may have changed/been scheduled already.
 			pod = &api.Pod{}
 			err := factory.Client.Get().Namespace(podNamespace).Resource("pods").Name(podID).Do().Into(pod)
```
Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: Can you change the call to Add below to AddIfNotPresent?

File: plugin/pkg/scheduler/factory/factory.go [hunk 4]
```diff
@@ -274,9 +276,16 @@ func (factory *ConfigFactory) makeDefaultErrorFunc(backoff *podBackoff, podQueue
 		// Note that this is extremely rudimentary and we need a more real error handling path.
 		go func() {
 			defer util.HandleCrash()
-			podID := pod.Name
+			podID := pod.Namespace + "/" + pod.Name
```
Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: Thanks for fixing this.

File: plugin/pkg/scheduler/factory/factory.go (hunk 1 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 2 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 3 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 4 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: ack.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

thanks
--brendan

**Comment by quinton-hoole** [comment]
@brendandburns 
--- FAIL: TestE2E (607.49s)
21:42:07 [31] 	e2e_test.go:132: Error waiting for all pods to be running and ready: Not all pods in namespace 'kube-system' running and ready within 10m0s

File: plugin/pkg/scheduler/factory/factory.go (hunk 4 above)
Comment (quinton-hoole) [inline plugin/pkg/scheduler/factory/factory.go]: Use NamespacedName perhaps, for general hygiene?

Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: To be super pedantic, there's a `cache.MetaNamespaceKeyFunc` which is what all the fifo/cache stuff uses.

File: plugin/pkg/scheduler/factory/factory.go [hunk 5]
```diff
@@ -334,9 +341,50 @@ func (realClock) Now() time.Time {
 	return time.Now()
 }
 
+// backoffEntry is single threaded.  in particular, it only allows a single action to be waiting on backoff at a time.
+// It is expected that all users will only use the public TryWait(...) method
```
Comment (lavalamp) [inline plugin/pkg/scheduler/factory/factory.go]: might be worth adding "not safe to copy"

**Comment by lavalamp** [comment]
This LGTM when tests pass, nits are optional

File: plugin/pkg/scheduler/factory/factory.go (hunk 4 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: What is the advantage of that method here?  It seems to just return the Namespace in an array...

File: plugin/pkg/scheduler/factory/factory.go (hunk 5 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: done.

File: plugin/pkg/scheduler/factory/factory.go (hunk 4 above)
Comment (brendandburns) [inline plugin/pkg/scheduler/factory/factory.go]: side channel conversation w/ @lavalamp indicates it's not required.

@quinton-hoole I don't quite follow your suggestion.

**Comment by brendandburns** [comment]
Comments addressed (mostly)

ptal.

--brendan

**Comment by brendandburns** [comment]
@quinton-hoole I think your comment is addressed now.

ptal

thanks!
--brendan

**Comment by quinton-hoole** [comment]
@brendandburns did you find a bug in your code, causing the e2e tests to fail?  I'll take a look as soon as the test results come back from Jenkins.

**Pull request merged**